tutorials/application_design/c++11/publisher.cxx

```diff
@@ -0,0 +1,133 @@
+//
+// (c) 2024 Copyright, Real-Time Innovations, Inc.  All rights reserved.
+//
+//  RTI grants Licensee a license to use, modify, compile, and create derivative
+//  works of the Software.  Licensee has the right to distribute object form
+//  only for use with RTI products.  The Software is provided "as is", with no
+//  warranty of any type, including any warranty for fitness for any purpose.
+//  RTI is under no obligation to maintain or support the Software.  RTI shall
+//  not be liable for any incidental or consequential damages arising out of the
+//  use or inability to use the software.
+//
+
+#include <iostream>
+#include <sstream>
+#include <thread>
+
+#include <rti/rti.hpp>
+#include "VehicleModeling.hpp"
+#include "common.hpp"
+
+struct PublisherSimulation {
+    explicit PublisherSimulation(
+            dds::pub::DataWriter<VehicleMetrics> metrics_writer,
+            dds::pub::DataWriter<VehicleTransit> transit_writer)
+            : metrics_writer_(metrics_writer),
+              transit_writer_(transit_writer),
+              vehicle_vin_(utils::new_vin()),
+              vehicle_fuel_(100.0),
+              vehicle_route_(utils::new_route()),
+              vehicle_position_(vehicle_route_[0])
+    {
+    }
+
+    bool has_ended() const
+    {
+        return is_out_of_fuel();
+    }
+
+    void run();
+
+    friend std::string utils::to_string<>(PublisherSimulation const &sim);
```

Comment:
I chose this approach because I didn't want to add accessor methods nor bind the function as a method, so that the definition remains small and straightforward.

I can make this function a bound method if you think that's needed.

Reply:
Yes, I think it makes more sense to define `to_string` as a method and call it from the free function.